odl/solvers/smooth/newton.py

```diff
@@ -0,0 +1,380 @@
+                hessinv_estimate=None, callback=None):
+    """Quasi-Newton BFGS method to minimize a differentiable function.
+
+    Can use either the regular BFGS method, or the limited memory BFGS method.
+
+    Notes
+    -----
+    This is a general and optimized implementation of a quasi-Newton
+    method with BFGS update for solving a general unconstrained
+    optimization problem
+
+        :math:`\min f(x)`
+
+    for a differentiable function
+    :math:`f: \mathcal{X}\\to \mathbb{R}` on a Hilbert space
+    :math:`\mathcal{X}`. It does so by finding a zero of the gradient
+
+        :math:`\\nabla f: \mathcal{X} \\to \mathcal{X}`.
+
+    The QN method is an approximate Newton method, where the Hessian
+    is approximated and gradually updated in each step. This
+    implementation uses the rank-one BFGS update schema where the
+    inverse of the Hessian is recalculated in each iteration.
+
+    The algorithm is described in [GNS2009]_, Section 12.3 and in the
+    `BFGS Wikipedia article
+    <https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93\
+Goldfarb%E2%80%93Shanno_algorithm>`_
+
+    Parameters
+    ----------
+    f : `Functional`
+        Functional with ``f.gradient``
+    x : ``f.domain`` element
+        Starting point of the iteration
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, uses it as a
+        fixed step length.
+    maxiter : int, optional
+        Maximum number of iterations.
+        ``tol``.
+    tol : float, optional
+        Tolerance that should be used for terminating the iteration.
+    maxcor : int, optional
+        Maximum number of correction factors to store. If ``None``, the method
+        is the regular BFGS method. If an integer, the method becomes the
+        Limited Memory BFGS method.
+    hessinv_estimate : `Operator`, optional
+        Initial estimate of the inverse of the Hessian operator. Needs to be an
+        operator from ``f.domain`` to ``f.domain``.
+        Default: Identity on ``f.domain``
+    callback : `callable`, optional
+        Object executing code per iteration, e.g. plotting each iterate.
+    """
+    grad = f.gradient
+    if x not in grad.domain:
+        raise TypeError('`x` {!r} is not in the domain of `grad` {!r}'
+                        ''.format(x, grad.domain))
+
+    if not callable(line_search):
+        line_search = ConstantLineSearch(line_search)
+
+    ys = []
+    ss = []
+
+    grad_x = grad(x)
+    for _ in range(maxiter):
+        # Determine a stepsize using line search
+        search_dir = -_bfgs_multiply(ys, ss, grad_x, hessinv_estimate)
+        dir_deriv = search_dir.inner(grad_x)
+        if np.abs(dir_deriv) < tol:
+            return  # we found an optimum
+        step = line_search(x, direction=search_dir, dir_derivative=dir_deriv)
+
+        # Update x
+        x_update = search_dir
+        x_update *= step
+        x += x_update
```

Comment:
No, I use the scaled guy later so I think this may be fast enough, using lincomb could shave of a very small time diff but would add a non-obvious step term a few lines further down.